src/tools/tidy/src/error_codes.rs

```diff
@@ -0,0 +1,397 @@
+            found_code_example = true;
+            found_proper_doctest = true;
+        }
+    }
+
+    (found_code_example, found_proper_doctest, emit_ignore_warning, emit_no_longer_warning)
+}
+
+// Stage 3: Checks that each error code has a UI test in the correct directory
+fn check_error_codes_tests(root_path: &Path, error_codes: &[String], errors: &mut Vec<String>) {
+    let tests_path = root_path.join(Path::new(ERROR_TESTS_PATH));
+
+    // Some warning counters, this whole thing is clunky but'll be removed eventually.
+    let mut no_ui_test = 0;
+    let mut no_error_code_in_test = 0;
+
+    for code in error_codes {
+        let test_path = tests_path.join(format!("{}.stderr", code));
+
+        if !test_path.exists() && !IGNORE_UI_TEST_CHECK.contains(&code.as_str()) {
+            no_ui_test += 1;
+            continue;
+        }
+        if IGNORE_UI_TEST_CHECK.contains(&code.as_str()) {
+            if test_path.exists() {
+                errors.push(format!(
+                    "Error code `{code}` has a UI test, it shouldn't be listed in `EXEMPTED_FROM_TEST`!"
+                ));
+            }
+            continue;
+        }
+
+        let file = match fs::read_to_string(test_path) {
+            Ok(file) => file,
+            Err(err) => {
+                println!(
+                    "WARNING: Failed to read UI test file for `{code}` but the file exists. The test is assumed to work:\n{err}"
+                );
+                continue;
+            }
+        };
+
+        let mut found_code = false;
+
+        for line in file.lines() {
+            let s = line.trim();
+            // Assuming the line starts with `error[E`, we can substring the error code out.
+            if s.starts_with("error[E") {
+                if &s[6..11] == code {
+                    found_code = true;
+                    break;
+                }
+            };
+        }
+
+        if !found_code {
+            no_error_code_in_test += 1;
+        }
+    }
+
+    if no_error_code_in_test > 0 {
+        println!(
+            "WARNING: {no_error_code_in_test} error codes have a UI test file, but don't contain their own error code!"
+        );
+    }
+
+    if no_ui_test > 0 {
+        println!(
+            "WARNING: {no_ui_test} error codes need to have at least one UI test in the `src/test/ui/error-codes/` directory`! \
+            This *will* become a hard error."
+        );
+    }
+}
+
+/// Stage 4: Search `compiler/` and ensure that every error code is actually used by the compiler and that no undocumented error codes exist.
+fn check_error_codes_used(
+    search_paths: &[&Path],
+    error_codes: &[String],
+    errors: &mut Vec<String>,
+    no_longer_emitted: &[String],
+) {
+    // We want error codes which match the following cases:
+    //
+    // * foo(a, E0111, a)
+    // * foo(a, E0111)
+    // * foo(E0111, a)
+    // * #[error = "E0111"]
+    let regex = Regex::new(r#"[(,"\s](E\d{4})[,)"]"#).unwrap();
```

Comment:
regex compile can be moved to lazy_static/once_cell?

Reply:
It seems like it's compiled only once, that shouldn't matter.

Reply:
It is, this was a change from how it was previously done I believe.